dbos-rules.ts

```diff
@@ -0,0 +1,442 @@
+
+type GlobalTools = {eslintContext: EslintContext, parserServices: ParserServicesWithTypeInformation, typeChecker: TypeChecker};
+let GLOBAL_TOOLS: GlobalTools | undefined = undefined;
+
+// These included `Transaction` and `TransactionContext` respectively before!
+const DETERMINISTIC_DECORATORS = new Set(["Workflow"]);
+const TYPES_YOU_CAN_AWAIT_UPON_IN_DETERERMINISTIC_FUNCTIONS = new Set(["WorkflowContext"]);
+const ERROR_MESSAGES = makeErrorMessageSet();
+
+////////// This is the set of error messages that can be emitted
+
+function makeErrorMessageSet(): Map<string, string> {
+  const makeDateMessage = (bannedCall: string) => `Calling ${bannedCall} is banned \
+(consider using \`@dbos-inc/communicator-datetime\` for consistency and testability)`;
+
+  const bcryptMessage = "Avoid using `bcrypt`, which contains native code. Instead, use `bcryptjs`. \
+Also, some `bcrypt` functions generate random data and should only be called from communicators";
+
+  const validTypeSetString = [...TYPES_YOU_CAN_AWAIT_UPON_IN_DETERERMINISTIC_FUNCTIONS].map((name) => `\`${name}\``).join(", ");
+
+  // The keys are the ids, and the values are the messages themselves
+  return new Map([
+    ["globalModification", "This is a global modification relative to the workflow declaration"],
+    ["awaitingOnNotAllowedType", `This function should not await with the current leftmost type (allowed set: \{${validTypeSetString}\})`],
+    ["Date", makeDateMessage("`Date()` or `new Date()`")],
+    ["Date.now", makeDateMessage("`Date.now()`")],
+    ["Math.random", "Avoid calling `Math.random()` directly; it can lead to non-reproducible behavior. See `@dbos-inc/communicator-random`"],
+    ["setTimeout", "Avoid calling `setTimeout()` directly; it can lead to undesired behavior when debugging"],
+    ["bcrypt.hash", bcryptMessage],
+    ["bcrypt.compare", bcryptMessage]
+  ]);
+}
+
+//////////
+
+/* Typically, awaiting on something in a workflow function is not allowed,
+since awaiting usually indicates IO, which may be nondeterministic. The only exception
+is awaiting on a call hinging on a `WorkflowContext`, e.g. for some code like this
+(where `ctxt` is a `WorkflowContext` object):
+
+`const user = await ctxt.client<User>('users').select("password").where({ username }).first();`
+
+But there's a common pattern of awaiting upon a function that doesn't have a leftmost `ctxt` there,
+but rather upon a function where you just pass that context in as a parameter. Some hypothetical code
+for that would look like this:
+
+`const user = await getUser(ctxt, username);`
+
+While this seems nondeterministic, it's likely to be deterministic, since the `getUser` function
+probably just does the snippet above, but in an abstracted manner (so `getUser` would be a helper function).
+So, setting this flag means that determinism warnings will be disabled for awaits in this situation. */
+const ignoreAwaitsForCallsWithAContextParam = true;
```

Comment:
Yeah, I think this is the right behavior, limit false positives by assuming deterministic any function that takes a context as input.